glib/src/subclass/signal.rs

```diff
@@ -68,9 +68,103 @@ impl fmt::Debug for SignalInvocationHint {
     }
 }
 
+/// In-depth information of a specific signal
+#[repr(transparent)]
+pub struct SignalQuery(gobject_ffi::GSignalQuery);
+
+impl SignalQuery {
+    pub fn signal_name(&self) -> crate::GString {
+        unsafe { from_glib_none(self.0.signal_name) }
+    }
+
+    pub fn signal_id(&self) -> SignalId {
+        unsafe { SignalId::new(from_glib(self.0.itype), self.0.signal_id) }
+    }
+
+    pub fn type_(&self) -> Type {
+        unsafe { from_glib(self.0.itype) }
+    }
+
+    pub fn flags(&self) -> SignalFlags {
+        unsafe { from_glib(self.0.signal_flags) }
+    }
+
+    pub fn return_type(&self) -> Type {
+        unsafe { from_glib(self.0.return_type) }
+    }
+
+    pub fn n_params(&self) -> u32 {
+        self.0.n_params
+    }
+
+    pub fn param_types(&self) -> Vec<Type> {
+        unsafe {
+            let types = self.0.param_types;
+            FromGlibContainerAsVec::from_glib_none_num_as_vec(types, self.n_params() as usize)
+        }
+    }
+}
+
+impl fmt::Debug for SignalQuery {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("SignalQuery")
+            .field("signal_name", &self.signal_name())
+            .field("type", &self.type_())
+            .field("flags", &self.flags())
+            .field("return_type", &self.return_type())
+            .field("param_types", &self.param_types())
+            .finish()
+    }
+}
 /// Signal ID.
 #[derive(Debug, Clone, Copy)]
-pub struct SignalId(pub(super) Type, pub(super) u32);
+pub struct SignalId(Type, u32);
+
+impl SignalId {
+    unsafe fn new(type_: Type, id: u32) -> Self {
```

Comment:
Also as the type is only for validation, maybe we shouldn't include it? OTOH checking the type in `emit()` wastes the performance advantage of using the signal id over the name so...

Reply:
I don't know much about the signals internals, so I will just go ahead and remove the type checking from emit

Reply:
That should be OK, it just causes a critical warning then if you emit the signal on the wrong type

Reply:
Basically, it's less bad than disconnecting a signal handler from the wrong object :)

Reply:
Alright, i guess that's fine to be handled by the api users :)